subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/hived_algorithm.go

```diff
@@ -529,242 +600,231 @@ func (h *HivedAlgorithm) findVirtualCell(
+				FailedNodeReasons: map[string]string{},
+			},
 		}
 	} else {
-		setPriority(physicalCell, opportunisticPriority, unlimitedLevel)
-	}
-}
-
-// confirmReleasedCell releases an allocated cell by destroying cell bindings and reset priorities.
-func confirmReleasedCell(c *PhysicalCell) {
-	var preassignedPhysical *PhysicalCell
-	if vc := c.GetVirtualCell(); vc != nil {
-		// destroy bindings
-		boundVirtual := vc
-		for !boundVirtual.GetPhysicalCell().IsReserved() {
-			boundPhysical := boundVirtual.GetPhysicalCell()
-			klog.Infof("Cells unbound: %v and %v", boundVirtual.GetName(), boundPhysical.GetName())
-			boundPhysical.SetVirtualCell(nil)
-			boundVirtual.SetPhysicalCell(nil)
-			if boundVirtual.GetParent() == nil {
-				break
-			} else {
-				unbindParent := true
-				for _, cc := range boundVirtual.GetParent().GetChildren() {
-					if child := cc.(*VirtualCell); child.GetPhysicalCell() != nil {
-						unbindParent = false
-						break
-					}
-				}
-				if !unbindParent {
-					break
-				}
-				boundVirtual = boundVirtual.GetParent().(*VirtualCell)
+		affinityGroupBindInfo, selectedNode, selectedGpuIndices := generateAffinityGroupBindInfo(
+			groupPhysicalPlacement, groupVirtualPlacement, currentGpuNum, currentPodIndex)
+		chain := string(groupPhysicalPlacement[currentGpuNum][currentPodIndex][0].GetChain())
+		if !common.StringsContains(suggestedNodes, selectedNode) {
+			panic(fmt.Sprintf("[%v]: node %v picked by algorithm but not in K8S candidates",
+				internal.Key(pod), selectedNode))
+		}
+		// collect preemption victims
+		preemptionVictims := common.NewSet()
+		for _, gpu := range groupPhysicalPlacement[currentGpuNum][currentPodIndex] {
+			pGpu := gpu.(*PhysicalCell)
+			if pGpu.HasPod() {
+				preemptionVictims.Add(pGpu.GetPods()[0])
 			}
 		}
-		// reset priority
-		setSelfAndParentPriority(vc, freePriority, unlimitedLevel)
-		preassignedPhysical = vc.GetPreAssignedCell().GetPhysicalCell()
-	} else {
-		// check if the cell is inside a physical cell of a preassigned virtual cell
-		for cc := c; ; {
-			if boundVirtual := cc.GetVirtualCell(); boundVirtual != nil {
-				preassignedPhysical = boundVirtual.GetPreAssignedCell().GetPhysicalCell()
-				break
+		if preemptionVictims.IsEmpty() {
+			klog.Infof("[%v]: scheduled to node %v, GPUs %v",
+				internal.Key(pod), selectedNode, selectedGpuIndices)
+			return internal.PodScheduleResult{
+				PodBindInfo: &api.PodBindInfo{
+					Node:                  selectedNode,
+					GpuIsolation:          selectedGpuIndices,
+					CellChain:             chain,
+					AffinityGroupBindInfo: affinityGroupBindInfo,
+				},
 			}
-			if cc.GetParent() != nil {
-				cc = cc.GetParent().(*PhysicalCell)
-			} else {
-				break
+		} else {
+			var victimPods []*core.Pod
+			for v := range preemptionVictims.Items() {
+				victimPods = append(victimPods, v.(*core.Pod))
+			}
+			return internal.PodScheduleResult{
+				PodPreemptInfo: &internal.PodPreemptInfo{VictimPods: victimPods},
 			}
```

Comment:
Wait until all victims in the group cleared

Reply:
And tell to preempt only if the whole group has sufficient allocated lower priority pods.

---
In reply to: [335258080](https://github.com/microsoft/pai/pull/3684#discussion_r335258080) [](ancestors = 335258080)

Reply:
The principle is that we are scheduling the whole group (as a gang) instead of the individual pod inside it.

---
In reply to: [335270249](https://github.com/microsoft/pai/pull/3684#discussion_r335270249) [](ancestors = 335270249,335258080)